gh-387 - scaffold submission service

abwilson23 opened the pull request:
- Create initial implementation of the overall submission process flow including: interfaces, main submission control flow, generic step stub and basic testing. 
- slight start to the select project step just to have a step to test. 

closes #387

abwilson23 committed:
gh-387 - initial scaffolding

- add resources, submission-service, submission-state, step interface.

abwilson23 committed:
gh-387 - complete initial scaffolding of submission process.

- implement stateService, basic testing, overall process flow
- some overlap with the select-project task.

abwilson23 committed:
gh-387-scaffold-submission-service - add license headers.

NigelPalmer reviewed (changes requested):
This is looking like a good start, well done!

Now we have this in place, I would like to take this task a little further to prove that the submission service will behave in the way that we would like. To that end, please update the task with the following functionality:

1. Change the "Select Project Dialog" to be a dialog with a button on it that says "Click to complete step 1". When clicked the specification description should be updated with the text "Submission Step 1 Complete". 
2. Update `SelectProjectStep.run` so that the stepResult includes a string called `whatStep1Did`. Set `whatStep1Did` to the message "Step 1 ran this time"
3. Update `SelectProjectStep.isRequired` to check the specification description:
If the description includes the text "Submission Step 1 Complete",  set `isRequired` to false and set `whatStep1Did` to "Step 1 was not required to be run".
If the description _does not_ include the text "Submission Step 1 Complete", set `isRequired` to true 
4. Create a Step 2. Step 2 should simply show a dialog that displays the  `whatStep1Did` message.

Regarding code review, nothing major has leapt out yet, but I would rather carry out a more thorough code rerview once the above is complete.

Apologies for moving the goal posts, but I think this will be worth doing.

abwilson23 committed:
gh-387-scaffold-submission-service - prove multiple steps work.

NigelPalmer commented on src/app/data-explorer/specification-submission/submission-steps/select-project.step.ts:
```diff
@@ -0,0 +1,87 @@
+SPDX-License-Identifier: Apache-2.0
+*/
+import { Injectable } from '@angular/core';
+import { Observable, map, of, switchMap, tap } from 'rxjs';
```
switchMap is not used. This will fail linting.

abwilson23 replied:
Resolved.

NigelPalmer commented on src/app/data-explorer/specification-submission/submission-steps/select-project.step.ts:
```diff
@@ -0,0 +1,87 @@
+          ? 'SelectProjectStep is required.'
+          : 'Step 1 was not required to be run.',
+      },
+      isRequired: isRequired,
```
'Expected property shorthand' error is being shown in my Visual Studio Code. Please change to just be isRequired.

abwilson23 replied:
Resolved.

NigelPalmer commented on src/app/data-explorer/specification-submission/submission-steps/select-project.step.ts:
```diff
@@ -0,0 +1,87 @@
+          () =>
+            ({
+              result: { whatStep1Did: 'Step 1 ran this time.' },
+              isRequired: false,
```
Why is isRequired being set to false here? This is confusing to follow. What is this isRequired even being used for after this point? Could we not just default isRequired to false in the StepResult and therefore not set the value here?

abwilson23 replied:
I went with removing it altogether as only the result is used in the main submit method to save any new data to the state.

NigelPalmer commented on src/app/data-explorer/specification-submission/submission.resource.ts:
```diff
@@ -0,0 +1,64 @@
+  | 'AttachSqlFileToDataRequest'
+  | 'GeneratePdfOfDataRequest'
+  | 'AttachPdfToDataRequest'
+  | 'SaveDataRequest';
```
The idea of StepName was that it is something that would be displayed to the user when a submission is running. Therefore these should be sentence case rather than looking code like. e.g:
'Select project'
'Get Data Request'
'Generate SQL file'
etc

abwilson23 replied:
Changed.

NigelPalmer reviewed (changes requested):
This was not quite what I had in mind, but the implementation clearly shows the step process working. Provided I type in "Submission step 1 complete." into the dialog, then the subsequent steps will fast track to step 2. If anything else is typed, then step 1 is run again. This is good and shows everything working as it should.

There are a couple of code observations that need addressing.

abwilson23 committed:
gh-387-scaffold-submission-service - address pr comments.

- use sentence case for step names
- remove unused property in returned stepResult

NigelPalmer reviewed (approved):
This framework is now ready to be populated with steps. This is all working well and a simple and flexible framework.
